cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java

```diff
@@ -114,12 +127,19 @@ public TransformRunner(MapReduceTaskContext<Object, Object> context,
                                                      PipelinePhase pipelinePhase,
                                                      Configuration hConf) {
     Set<StageInfo> aggregators = pipelinePhase.getStagesOfType(BatchAggregator.PLUGIN_TYPE);
+    Set<StageInfo> joiners = pipelinePhase.getStagesOfType(BatchJoiner.PLUGIN_TYPE);
     if (!aggregators.isEmpty()) {
       String aggregatorName = aggregators.iterator().next().getName();
       if (pipelinePhase.getSinks().contains(aggregatorName)) {
         return new SingleOutputWriter<>(context);
       }
     }
+    if (!joiners.isEmpty()) {
```

Comment:
duplication here can be removed too by doing a union of the sets.  It may also be worthwhile to add a `getStagesOfTypes(String ... types)` method to PipelinePhase.